Maflipulator Code

Quinn-5 opened the pull request:
Adds base code for the manipulator flipping and  adjustment mechanism, with some commands to flip, adjust position, and avoid collisions with the robot itself.

DavidDesfosse committed:
Lift and Maflipulator Classes now Exist

Made the lift and manipulator classes, made a command for manual control of lift, edited the robot map in order to add lift and maflipulator classes

DavidDesfosse committed:
Added commands

DavidDesfosse committed:
Add enum functionality

now tests which side the maflipularor is on and flips in the opposite direction accordingly

Quinn-5 committed:
Fixed restrict roatate

edited restrict rotate so that it worked for both sides, gave restrict rotate a name

Quinn-5 committed:
Deleted lift code

It's gone now.... very sad, so much work went into it and now the lift has died

msoucy commented on src/main/java/frc/robot/maps/PracticeBot.java:
```diff
@@ -45,4 +49,47 @@ public DigitalInput getGamepieceLimitSwitch() {
+
+            @Override
+            public SendableSpeedController getLiftMotor() {
+                return SendableSpeedController.wrap(new WPI_VictorSPX(7));
```
Use `new MockSpeedController()` instead, for now.

msoucy commented on src/main/java/frc/robot/maps/PracticeBot.java:
```diff
@@ -45,4 +49,47 @@ public DigitalInput getGamepieceLimitSwitch() {
+
+            @Override
+            public SendableSpeedController getFlipMotor() {
+                return SendableSpeedController.wrap(new WPI_VictorSPX(4));
```
Use `new MockSpeedController()` instead, for now.

msoucy commented on src/main/java/frc/robot/subsystems/Maflipulator.java:
```diff
@@ -0,0 +1,143 @@
+                return false;
+            }
+
+            @Override
```
Delete this function and `interrupted`, since there's nothing special about them

msoucy commented on src/main/java/frc/robot/subsystems/Maflipulator.java:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    public Command restrictRotate() {
+        // The command is named "Sample Command" and requires this subsystem.
```
This comment is inaccurate

msoucy commented on src/main/java/frc/robot/subsystems/Maflipulator.java:
```diff
@@ -0,0 +1,143 @@
+                if (currentPosition == MaflipulatorSide.kFront) {
+                    flipMotor.set(1);
+                }
+                if (currentPosition == MaflipulatorSide.kBack) {
```
Why not just use `else` ?

Quinn-5 committed:
Fixed problems for pull request

Fixed all issues with pull request

bot190 commented on src/main/java/frc/robot/RobotMap.java:
```diff
@@ -32,4 +34,25 @@
 
+    LiftMap getLiftMap();
+
+    interface LiftMap {
```
This should be removed from this PR as we already have a lift map.

bot190 commented on src/main/java/frc/robot/RobotMap.java:
```diff
@@ -32,4 +34,25 @@
+
+        SendableSpeedController getFlipMotor();
+
+        Potentiometer getManipulatorPot();
```
`getMaflipulatorPot`, Please keep the naming consistent as the manipulator is a separate subsystem. (No this probably isn't the best choice)

bot190 commented on src/main/java/frc/robot/maps/PracticeBot.java:
```diff
@@ -45,4 +50,47 @@ public DigitalInput getGamepieceLimitSwitch() {
     // Fill in any methods from the interface here.
+
+    @Override
+    public LiftMap getLiftMap() {
```
As above, please remove this

bot190 commented on src/main/java/frc/robot/subsystems/Maflipulator.java:
```diff
@@ -0,0 +1,129 @@
+    private MaflipulatorSide currentPosition;
+
+    private SendableSpeedController flipMotor;
+    private Potentiometer manipulatorPot;
```
Should be `maflipulatorPot`

bot190 commented on src/main/java/frc/robot/subsystems/Maflipulator.java:
```diff
@@ -0,0 +1,129 @@
+    @Override
+    public void initDefaultCommand() {
+        // Set the default command for a subsystem here.
+        setDefaultCommand(manualFlip());
```
Lets set the default command to `restrictRotate()`

bot190 commented on src/main/java/frc/robot/subsystems/Maflipulator.java:
```diff
@@ -0,0 +1,129 @@
+            protected void execute() {
+                double flipSpeed = Robot.coPilot.getY(Hand.kRight);
+                if (currentPosition == MaflipulatorSide.kFront) {
+                    if (flipSpeed > 0 && manipulatorPot.get() >= 180) {
```
All of these angles should be defined at the top of the subsystem as `private final const double`

bot190 commented on src/main/java/frc/robot/subsystems/Maflipulator.java:
```diff
@@ -0,0 +1,129 @@
+                }
+
+                if (currentPosition == MaflipulatorSide.kBack) {
+                    if (flipSpeed > 0 && manipulatorPot.get() <= 180) {
```
Keep in mind that this is not really going to be 180, and should be different from above.

bot190 commented on src/main/java/frc/robot/subsystems/Maflipulator.java:
```diff
@@ -0,0 +1,129 @@
+            @Override
+            protected void initialize() {
+                if (currentPosition == MaflipulatorSide.kFront) {
+                    flipMotor.set(1);
```
Should probably be a constant we can easily change

bot190 commented on src/main/java/frc/robot/subsystems/Maflipulator.java:
```diff
@@ -0,0 +1,129 @@
+                if (currentPosition == MaflipulatorSide.kFront && manipulatorPot.get() >= 270) {
+                    return true;
+                }
+                if (currentPosition == MaflipulatorSide.kBack && manipulatorPot.get() <= 90) {
```
Could just be ` else if (manipulatorPot.get() <= 90)` since there's only two possible sides. (We could probably just use a bool here. But enums are fun!

bot190 reviewed (changes requested):
We should also look into using PID to control the angle of the maflipulator. This is actually a really good use case for it.

DavidDesfosse committed:
Fixed things, it actually builds now

Fixed robot map, also created constants for the angle restrictions and flip motor speed. Fully got rid of lift

DavidDesfosse committed:
Added PID

Added PID to maflipulator

DavidDesfosse committed:
Made restrict function

Made a function that restricts the maflipulator from moving past a certain angle

Quinn-5 committed:
Change Maflipulator potentiometer in Practicebot map

Change AnalogPotentiometer to MockPotentiometer

Quinn-5 committed:
Change "Move to Position" into a PIDCommand

Chnages the Maflipulator's "Move to Position" Command into a PIDCommand

Quinn-5 committed:
Determine which side manipulator is on at start

Code now determines which side the manipulator is on at start

Quinn-5 committed:
PIDFlip Command

Add a command that will maybe possibly work to send the manipulator to one of four wanted positions, depending on where it is at the start

Quinn-5 committed:
Split PIDFlip command

Split the command into two - one per each position on either side

Quinn-5 committed:
Implement PID into Flip command

Flip now uses PID

Quinn-5 committed:
Fix PID positioning command names

The commands had swapped names, now they don't

msoucy commented on src/main/java/frc/robot/subsystems/Maflipulator.java:
```diff
@@ -147,18 +139,35 @@ protected boolean isFinished() {
+        return new InstantCommand("PID Flip", this, () -> {
+            if (currentPosition == MaflipulatorSide.kFront) {
+                if (anglePot.get() < (FRONT_LOWER_ANGLE + 90) / 2) {
+                    moveToPosition(FRONT_UPPER_ANGLE);
```
This and the line below it create commands, but don't actually start them.

msoucy replied:
Never mind, Github led me astray

billcalabro commented on src/main/java/frc/robot/subsystems/Maflipulator.java:
```diff
@@ -0,0 +1,193 @@
+
+            Command moveCommand;
+            if (currentPosition == MaflipulatorSide.kFront) {
+                moveCommand = moveToPosition(270);
```
Could we define these are variables up top? Just so if we need to adjust them they are easy to find.

billcalabro replied:
Similar to the BACK_LOWER_ANGLE definitions. Basically I don't want to see any random numbers mixed in this deep in the code (90, 270, etc.)

msoucy commented on src/main/java/frc/robot/subsystems/Maflipulator.java:
```diff
@@ -0,0 +1,193 @@
+        setDefaultCommand(restrictRotate());
+    }
+
+    public Command manualFlip() {
```
`manualFlip` doesn't seem to be used anymore, or serve a real purpose in favor of `restrictRotate`. Recommend deleting

msoucy commented on src/main/java/frc/robot/subsystems/Maflipulator.java:
```diff
@@ -0,0 +1,193 @@
+    private final static double BACK_LOWER_ANGLE = 290;
+    private final static double BACK_UPPER_ANGLE = 180;
+
+    private final static double FLIP_MOTOR_SPEED = 1;
```
This should probably be lower, at least for testing.

msoucy commented on src/main/java/frc/robot/subsystems/Maflipulator.java:
```diff
@@ -0,0 +1,193 @@
+
+            @Override
+            protected boolean isFinished() {
+                return anglePID.onTarget();
```
Look at adding a `&& Math.abs(angleCorrection) < 0.2` or similar, to make sure it stops at the right location

msoucy commented on src/main/java/frc/robot/subsystems/Maflipulator.java:
```diff
@@ -0,0 +1,193 @@
+                currentPosition = MaflipulatorSide.kFront;
+            }
+
+            Scheduler.getInstance().add(moveCommand);
```
I recommend using `moveCommand.start()` instead - it does some additional checks

Quinn-5 committed:
Replace Scoring position values with constants

Replace Scoring position values with constants defined  with other angle values at the top of the file

Quinn-5 committed:
Remove manualFlip and reduce flipspeed

This removes manualFlip in favor of restrictRotate, reduces the maximum flip motor speed, and now multiplies the controller input values in restrictRotate, so the speed should now actually be limited to FLIP_MOTOR_SPEED

Quinn-5 committed:
Replace Scheduler.getInstance

Replace Scheduler.getInstance().add(movecommand) ; with moveCommand.start();

Quinn-5 committed:
Add speed check before finishing MoveToPosition

Now only stops the command once the motors are slow enough, to reduce error

Quinn-5 committed:
Replace Flip command end position values with constants

Replace Flip command end position values with constants

bot190 reviewed (approved):
Awesome. This looks good.

bot190 commented:
Lauren will make Matt's suggested changes. We want to get this code in for integration.
